### PR TITLE
refactor: change visual regression testing API

### PR DESCRIPTION
**Pull request by kyubisation**
This PR changes the visual regression testing API to be more idiomatic for writing Mocha tests.

### CODE REVIEW

File: src/elements/core/testing/private/visual-regression-snapshot.ts [hunk 1]
```diff
@@ -20,61 +22,121 @@ function findElementCenter(snapshotElement: () => HTMLElement): [number, number]
   ];
 }
 
-export function testVisualDiff(snapshotElement: () => HTMLElement): void {
-  it('default', async function () {
-    await visualDiff(snapshotElement(), imageName(this.test!));
-  });
-}
+class VisualDiffSetup {
```
Comment (jeripeierSBB) [inline src/elements/core/testing/private/visual-regression-snapshot.ts:25]: Shouldn't we export the class?

File: src/elements/core/testing/private/visual-regression-snapshot.ts [hunk 2]
```diff
@@ -20,61 +22,121 @@ function findElementCenter(snapshotElement: () => HTMLElement): [number, number]
...
+      const context = new VisualDiffSetup();
+      await setup(context);
+      await applyViewportIfDefined(this.ctx['requestViewport']);
+      await visualDiff(context.snapshotElement, imageName(this.test!));
+    };
+  },
+};
+
+export const visualDiffFocus: VisualDiffState = {
+  name: 'default',
```
Comment (jeripeierSBB) [inline src/elements/core/testing/private/visual-regression-snapshot.ts:85]:
```suggestion
  name: 'focus',
```

File: src/elements/core/testing/private/visual-regression-snapshot.ts [hunk 3]
```diff
@@ -20,61 +22,121 @@ function findElementCenter(snapshotElement: () => HTMLElement): [number, number]
...
+      await applyViewportIfDefined(this.ctx['requestViewport']);
+      context.snapshotElement.focus();
+      await sendKeys({ press: 'Tab' });
+      await visualDiff(context.snapshotElement, imageName(this.test!));
+    };
+  },
+};
+
+export const visualDiffHover: VisualDiffState = {
+  name: 'default',
```
Comment (jeripeierSBB) [inline src/elements/core/testing/private/visual-regression-snapshot.ts:99]:
```suggestion
  name: 'hover',
```

File: src/elements/core/testing/private/visual-regression-snapshot.ts [hunk 4]
```diff
@@ -20,61 +22,121 @@ function findElementCenter(snapshotElement: () => HTMLElement): [number, number]
...
+  },
+};
+
+export const visualDiffActive: VisualDiffState = {
+  name: 'default',
+  with(setup: (setup: VisualDiffSetup) => void | Promise<void>): Mocha.Func {
+    return async function (this: Mocha.Context) {
+      const context = new VisualDiffSetup();
+      await setup(context);
+      await applyViewportIfDefined(this.ctx['requestViewport']);
```
Comment (jeripeierSBB) [inline src/elements/core/testing/private/visual-regression-snapshot.ts:123]: Those three lines are appearing multiple times. shouldn't we create a method or include it more globally?

File: src/elements/core/testing/private/visual-regression-snapshot.ts [hunk 5]
```diff
@@ -20,61 +22,121 @@ function findElementCenter(snapshotElement: () => HTMLElement): [number, number]
...
+        await visualDiff(context.snapshotElement, imageName(this.test!));
+      } finally {
+        await resetMouse();
+      }
+    };
+  },
+};
+
+export const visualDiffActive: VisualDiffState = {
+  name: 'default',
```
Comment (jeripeierSBB) [inline src/elements/core/testing/private/visual-regression-snapshot.ts:118]:
```suggestion
  name: 'active',
```

**Review by jeripeierSBB — CHANGES_REQUESTED** [review]
Thanks for improving the flow :)

File: src/elements/core/testing/private/visual-regression-snapshot.ts (hunk 1 above)
Comment (jeripeierSBB) [reply src/elements/core/testing/private/visual-regression-snapshot.ts:25]: if you need to import the types like when you want to pre define the setup new VisualDiffSetup()? probably not needed...

Comment (kyubisation) [reply src/elements/core/testing/private/visual-regression-snapshot.ts:25]: I think in that case we would need to reconsider the API again and if necessary just add the export keyword.
Or would you prefer to have it exported per default?

Comment (jeripeierSBB) [reply src/elements/core/testing/private/visual-regression-snapshot.ts:25]: nono, all ok :)

**Review by jeripeierSBB — APPROVED** [review]
LGTM :-). Could you please check the visual regression app, if everything still looks like expected? Only to be sure ;)

**Review by TomMenga — APPROVED** [review]
I like it 👍  LGTM

**Pull request merged**